R/batchmark.R

```diff
@@ -109,38 +100,37 @@ reduceBatchmarkResults = function(ids = NULL, keep.pred = TRUE, show.info = getM
   # registry and ids are asserted later
   requirePackages("batchtools", why = "batchmark", default.method = "load")
   assertFlag(keep.pred)
-
-  problems = batchtools::getProblemIds(reg)
-  algorithms = batchtools::getAlgorithmIds(reg)
-
-  result = replicate(length(problems), namedList(algorithms), simplify = FALSE)
-  names(result) = problems
-
-  for (prob in problems) {
-    exps = batchtools::findExperiments(prob.name = prob, ids = ids)
-    job = batchtools::makeJob(id = exps[1, ])
-    rin = job$instance$rin
-
-    if (nrow(exps) > 0) {
-      problem = job$problem
-
-      for (a in algorithms) {
-        # FIXME: Use prob.name + algo.name as soon as batchtools-0.9.3 hits CRAN
-        res = batchtools::reduceResultsList(batchtools::findExperiments(prob.pattern = stri_paste("^", prob, "$"),
-          algo.pattern = stri_paste("^", a, "$"), ids = ids))
-        models = !is.null(res[[1L]]$model)
-        lrn = problem$data$learner[[a]]
-        extract.this = getExtractor(lrn)
-        rs = mergeResampleResult(learner.id = a, task = problem$data$task, iter.results = res, measures = problem$data$measures,
-          rin = rin, keep.pred = keep.pred, models = models, show.info = show.info, runtime = NA, extract = extract.this)
-        rs$learner = lrn
-        result[[prob]][[a]] = addClasses(rs, "ResampleResult")
-        }
+  assertClass(reg, "ExperimentRegistry")
+
+  if (is.null(ids))
+    ids = batchtools::findDone(reg = reg)
+  if (NROW(ids) != nrow(batchtools::findExperiments(reg = reg)))
+    warning("Collecting results for a subset of jobs. The resulting BenchmarkResult may be misleading.")
+
+  problem = algorithm = NULL # for data.table's NSE
+  tab = batchtools::getJobPars(ids, flatten = FALSE, reg = reg)[, c("job.id", "problem", "algorithm")]
+  setkeyv(tab, cols = c("problem", "algorithm"), physical = FALSE)
+  result = namedList(tab[, unique(problem)])
+
+  for (prob in names(result)) {
+    algos = unique(tab[problem == prob], by = "algorithm")
+    data = batchtools::makeJob(id = algos$job.id[1L])$problem$data
+    result[[prob]] = namedList(algos$algorithm)
+
+    for (algo in names(result[[prob]])) {
+      res = batchtools::reduceResultsList(tab[problem == prob & algorithm == algo])
```

Comment:
Something I stumbled upon while playing around:
Why is `, reg = reg` not passed here? 
As I see it for now it only works with default registries?